src/parse.c

```diff
@@ -6218,14 +6222,23 @@ static tree_t p_subprogram_specification(void)
          type_set_result(type, p_type_mark());
       else {
          require_std(STD_19, "function knows return type") ;
+         tree_t t = tree_new(T_SUBTYPE_DECL);
```

Comment:
The previous version was actually correct: we don't want to create the `T_SUBTYPE_DECL` here but in `p_subprogram_body`. If you revert it to just creating the subtype `type_t` here, I can push something that does the second part - that's probably easier than trying to explain it in English :-).